packages/core/database/src/migrations/internal-migrations/5.0.0-03-document-discard-drafts.ts

```diff
@@ -0,0 +1,91 @@
+/* eslint-disable no-continue */
+import { async, contentTypes } from '@strapi/utils';
+
+import type { Database } from '../..';
+import type { Migration } from '../common';
+
+type DocumentVersion = { documentId: string; locale: string };
+
+/**
+ * Load a batch of versions to discard.
+ *
+ * Versions with only a draft version will be ignored.
+ * Only versions with a published version (which always have a draft version) will be discarded.
+ */
+async function* getBatchToDiscard({
+  db,
+  uid,
+  batchSize = 1000,
+}: {
+  db: Database;
+  uid: string;
+  batchSize?: number;
+}) {
+  let offset = 0;
+  let hasMore = true;
+
+  while (hasMore) {
+    // Look for the published entries to discard
+    const batch: DocumentVersion[] = await db
+      .queryBuilder(uid)
+      .select(['id', 'documentId', 'locale'])
+      .where({ publishedAt: { $ne: null } })
+      .limit(batchSize)
+      .offset(offset)
+      .orderBy('id')
+      .execute();
+
+    if (batch.length < batchSize) {
+      hasMore = false;
+    }
+
+    offset += batchSize;
+    yield batch;
+  }
+}
+
+const migrateUp = async (db: Database) => {
+  for (const meta of db.metadata.values()) {
+    const model = strapi.getModel(meta.uid);
+    const hasDP = contentTypes.hasDraftAndPublish(model);
+    if (!hasDP) {
+      continue;
+    }
+
+    const discardDraft = async (entry: DocumentVersion) => {
+      strapi
```

Comment:
I think we need to run that migration outsdie the DB package entirely now that it depends on the doc service it should not depend on the strapi global

We can add a `register` method in strapi.db.migrations.internal.register(migration) so that we can register migration to run from the outside to preserve decoupling

Reply:
will work on a proposal

Reply:
Initialy I exposed strapi.db.migrations.internal|user.register(migrationObj) but I realised I didn't need it for the document_id stuff